js/dataframe/cell.tsx

```diff
@@ -0,0 +1,279 @@
+      // If the new column index is out of bounds, quit
+      return;
+    }
+
+    setEditColumnIndex(newColumnIndex);
+  };
+  const onEnter = (e: ReactKeyboardEvent<HTMLInputElement>) => {
+    if (e.key !== "Enter") return;
+    // Prevent default behavior
+    e.preventDefault();
+
+    const hasShift = e.shiftKey;
+
+    const newRowIndex = editRowIndex! + (hasShift ? -1 : 1);
+    if (newRowIndex < 0 || newRowIndex >= maxRowSize) {
+      // If the new row index is out of bounds, quit
+      // attemptUpdate();
+      return;
+    }
+
+    setEditRowIndex(newRowIndex);
+  };
+
+  const onInputKeyDown = (e: ReactKeyboardEvent<HTMLInputElement>) => {
+    [onEsc, onEnter, onTab].forEach((fn) => fn(e));
+  };
+
+  const attemptUpdate = () => {
+    // Only update if the string form of the value has changed
+    if (`${initialValue}` === `${value}`) return;
+
+    setCellState(CellStateMap.EditSaving);
+    // console.log("Setting update count");
+    // setUpdateCount(updateCount + 1);
+    updateCellsData({
+      id,
+      cellInfos: [
+        {
+          rowIndex,
+          columnIndex,
+          value,
+          prev: initialValue,
+        },
+      ],
+      onSuccess: (values) => {
+        // console.log("success!", values);
+        setCellState(CellStateMap.EditSuccess);
+      },
+      onError: (err) => {
+        console.error("Error saving tabel cell value!", err);
+        setErrorTitle(String(err));
+        setCellState(CellStateMap.EditFailure);
+      },
+      columns,
+      setData,
+      setCellEditMap,
+    });
+  };
+
+  // When the input is blurred, we'll call our table meta's updateData function
+  // console.log("rendering cell", rowIndex, id, initialValue, value);
+  const onBlur = () => {
+    // console.log("on blur!", initialValue, value, e);
+    attemptUpdate();
+  };
+
+  // If the initialValue is changed external, sync it up with our state
+  useEffect(() => {
+    setValue(initialValue);
+  }, [initialValue, setValue]);
+
+  // Select the input when it becomes editable
+  useEffect(() => {
+    if (cellState !== CellStateMap.Editing) return;
+    if (!inputRef.current) return;
+
+    inputRef.current.focus();
+    inputRef.current.select();
+
+    // Setup global click listener to reset edit info
+    const onBodyClick = (e: MouseEvent) => {
+      if (e.target === inputRef.current) {
+        return;
+      }
+      resetEditInfo();
+    };
+    window.document
+      .querySelector("body")!
+      .addEventListener("click", onBodyClick);
+
+    // Tear down global click listener when we're done
+    return () => {
+      window.document
+        .querySelector("body")!
+        .removeEventListener("click", onBodyClick);
+    };
+  }, [cellState, resetEditInfo]);
+
+  // Reselect the input when it comes into view!
+  // (It could be scrolled out of view and then back into view)
+  function onFocus(e: ReactFocusEvent<HTMLInputElement>) {
```

Comment:
Restore cursor position and selection when scrolling back into view.

Reply:
Added TODO